condition/task.go

```diff
@@ -0,0 +1,181 @@
+package condition
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+)
+
+var (
+	errInvalidTaskJSON = errors.New("invalid Task JSON")
+	errTaskUpdate      = errors.New("invalid Task update")
+)
+
+const (
+	TaskVersion1 = "1.0"
+	// Task KV bucket name
+	TaskKVRepositoryBucket = "tasks"
+)
+
+func TaskKVRepositoryKey(facilityCode string, conditionKind Kind, serverID string) string {
+	return fmt.Sprintf("%s.%s.%s", facilityCode, conditionKind, serverID)
+}
+
+type Task[P, D any] struct {
+	// StructVersion indicates the Task object version and is used to determine Task  compatibility.
+	StructVersion string `json:"task_version"`
+
+	// Task unique identifier, this is set to the Condition identifier.
+	ID uuid.UUID `json:"id"`
+
+	// Kind is the type of Condition this Task is derived from
+	Kind Kind `json:"kind"`
+
+	// state is the state of the install
+	State State `json:"state"`
+
+	// status holds informational data on the state
+	Status StatusRecord `json:"status"`
+
+	// Data holds Condition Task specific data
+	Data D `json:"data,omitempty"`
+
+	// Parameters holds Condition specific parameters for this task
+	Parameters P `json:"parameters,omitempty"`
+
+	// Fault is a field to inject failures into a flasher task execution,
+	// this is set from the Condition only when the worker is run with fault-injection enabled.
+	Fault *Fault `json:"fault,omitempty"`
+
+	// FacilityCode identifies the facility this task is to be executed in.
+	FacilityCode string `json:"facility_code"`
+
+	// Asset holds attributes about the device under firmware install.
+	Asset *Asset `json:"asset,omitempty"`
+
+	// WorkerID is the identifier for the worker executing this task.
+	WorkerID string `json:"worker_id,omitempty"`
+
+	TraceID string `json:"traceID"`
+	SpanID  string `json:"spanID"`
+
+	CreatedAt   time.Time `json:"created_at,omitempty"`
+	UpdatedAt   time.Time `json:"updated_at,omitempty"`
+	CompletedAt time.Time `json:"completed_at,omitempty"`
+}
+
+// SetState implements the Task runner interface
+func (t *Task[P, D]) SetState(state State) {
+	t.State = state
+}
+
+func (t *Task[P, D]) Marshal() (json.RawMessage, error) {
+	return json.Marshal(t)
+}
+
+// TaskFromMessage converts a Task json.RawMessage object into its Task equivalent
+func TaskFromMessage(msg json.RawMessage) (*Task[any, any], error) {
+	t := &Task[any, any]{}
+	if err := json.Unmarshal(msg, t); err != nil {
+		return nil, errors.Wrap(errInvalidTaskJSON, err.Error())
+	}
+
+	return t, nil
+}
+
+// Update validates and applies the given changes to the Task object
```

Comment:
What is your risk profile for needing a validation method? I feel like ConditionOrchestrator shouldn't have a validation role here.

Reply:
The ConditionOrchestrator is not itself directly performing any Task Update and validation.

The risk profile is pretty much what is covered in the [tests](https://github.com/metal-toolbox/rivets/blob/18dacca3343b84b6e402b1af1d8a3e3d623fcbf1/condition/task_test.go#L92)

This `Task.Update` method provides common, consistent Task update path which is used by,
- The events controller library task publisher https://github.com/metal-toolbox/rivets/blob/041bab5c86554cb8a5015e2fcf9fa51aa2fc5b0c/events/controller/task.go#L210
- The Orchestrator API endpoint that receives [Task updates](https://github.com/metal-toolbox/conditionorc/blob/server-reservation/pkg/api/v1/routes/handlers_orchestrator.go#L961)